### ISSUE TITLE
Error when using union types with --noImplicitAny enabled

### ISSUE THREAD

**Issue by zhantongz** [Microsoft/TypeScript#6357]
*For example*:
There are some interfaces and a class:
```ts
interface Named {
    name: string;
    hello(para: Person | Pet): any;
}

interface Person {
    (para1: number, para2: string, para3: boolean): any;
}

interface Pet {
    (para1: number, para2: string): any;
}

class Test implements Named {
    name: string;
    hello(para: Person | Pet) {
        return true;
    }
    constructor(name: string) {
        this.name = name;
    }
}
```
With the following code
```ts
var test: Test = new Test("d");

test.hello((para1, para2, para3)=>{});
```
The compiler returns no error normally. However, if ```--noImplicitAny``` is enabled, the compiler returns
```
>$ tsc test --noImplicitAny
test.ts(26,13): error TS7006: Parameter 'para1' implicitly has an 'any' type.
test.ts(26,20): error TS7006: Parameter 'para2' implicitly has an 'any' type.
test.ts(26,27): error TS7006: Parameter 'para3' implicitly has an 'any' type.
```

If I change union type ```Person | Pet``` to just ```Person``` in ```Named``` and ```Test```, there would be no error even with ```--noImplicitAny``` enabled.

Is this an intended behaviour? It seems to me that using union types should not be trigger an implicit any error since union types are still type-checked when compiling.

**Comment by RyanCavanaugh** [comment]
Signatures of union types don't merge in a way that produces a signature which can be used for contextual typing.

This shouldn't be a problem, in theory. In your case, `Person` is a supertype (for purposes of which functions can be passed to `hello`) of `Pet`, so it should simply be written `hello(para: Person)` and you'll no longer have any implicit `any`s. The cases where this *doesn't* work are the cases where we wouldn't be able to provide a meaningful contextual type anyway (for example, if the parameter types of `Person` and `Pet` were not respectively identical).

**Comment by zhantongz** [comment]
@RyanCavanaugh Thank you for prompt reply.
I did use a (over-)simplified version to illustrate my issue.
My issue currently is actually related to DefinitelyTyped/DefinitelyTyped#7461.

In that case, the argument types are not aligned so there would be problem. I guess in my example it would be like
```ts
interface Pet {
    (para1: number, para2: string): any;
}
```

While the contextual type may not be meaningful in this case with union types, it is still checking types (for example, ```
test.hello((para1, para2, para3,para4)=>{});``` wouldn't pass the compiler since it satisfies neither ```Person``` nor ```Pet```). I think it would be a good idea to include an additional flag to suppress --noImplicitAny errors that are resulted from using union types.

**Comment by RyanCavanaugh** [comment]
"noImplicitAny errors that are resulted from using union types" is an oversimplication that skips an important step - the problem is that the function expression is not contextually typed by something with a signature.

Looking at your specific example
```ts
        interface ErrorRequestHandler {
            (err: any, req: Request, res: Response, next: Function): any;
        }

        interface RequestHandler {
            (req: Request, res: Response, next: Function): any;
        }

        // Later
        use(path: RegExp, handler: RequestHandler | ErrorRequestHandler): T;
```

If this is the actual way the code works, then `use` is checking `handler.length` to decide whether to pass `err, req, res, next` or `req, res, next`. In that case, you should separate this out into two separate overloads -- using a union type is not a correct model of the function's behavior.

```ts
        use(path: RegExp, handler: RequestHandler): T;
        use(path: RegExp, handler: ErrorRequestHandler): T;
```
Putting the shorter overload first will ensure that the correct overload gets selected.

**Comment by zhantongz** [comment]
@RyanCavanaugh Thank you for your help. I now understand my issue is not really about union types but contextual typing of functions with overloads.

The two separate overloads were used previously and resulted DefinitelyTyped/DefinitelyTyped#4212.

In my example, if I used
```ts
interface Named {
    name: string;
    hello(para: Pet): any;
    hello(para: Person): any;
}
```
and subsequently used ```test.hello((par1, par2, par3)=>{});```, there would be an error: ```test.ts(29,12): error TS2345: Argument of type '(par1: number, par2: string, par3: any) => void' is not assignable to parameter of type 'Person'.```.

However, this can be worked around easily with an additional typed variable, e.g.
```ts
var person: Person = (par1, par2, par3)=>{};
test.hello(person);
```

Again thank you for your assistance.

**Comment by RyanCavanaugh** [comment]
Happy to help! Thanks for contributing to DefinitelyTyped
